modules/foreign-state-oracle/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+			ReserveIdentifier = ReserveIdentifier,
+		>;
+
+		/// The foreign state oracle module id, keeps expired queries deposits
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Fee to be paid to oracles for servicing query
+		#[pallet::constant]
+		type QueryFee: Get<Balance>;
+
+		/// Fee for cancelling query
+		#[pallet::constant]
+		type CancelFee: Get<Balance>;
+
+		/// Timeout for query requests
+		#[pallet::constant]
+		type QueryDuration: Get<Self::BlockNumber>;
+
+		/// Dispatchable task that needs to be verified by oracle for dispatch
+		type VerifiableTask: Parameter
+			+ Dispatchable<Origin = <Self as Config>::Origin, PostInfo = PostDispatchInfo>
+			+ From<frame_system::Call<Self>>
+			+ GetDispatchInfo;
+
+		/// Origin that can dispatch calls that have been verified with foreign state
+		type OracleOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		/// Provides current blocknumber
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = Self::BlockNumber>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::origin]
+	pub type Origin = RawOrigin;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Active Query is created, under the index as the key
+		CreateActiveQuery { index: QueryIndex, expiry: T::BlockNumber },
+		/// Call is dispatched with data, includes the result of the extrinsic
+		CallDispatched { task_result: DispatchResult },
+		/// Query that has expired is removed from storage, includes index
+		StaleQueryRemoved { query_index: QueryIndex },
+		/// Query is canceled, includes index
+		QueryCanceled { index: QueryIndex },
+	}
+
+	/// Index of Queries, each query gets unique number
+	#[pallet::storage]
+	#[pallet::getter(fn query_index)]
+	pub(super) type QueryCounter<T: Config> = StorageValue<_, QueryIndex, ValueQuery>;
+
+	/// The tasks to be dispatched with data provideed by foreign state oracle
+	///
+	/// ActiveQuery: map QueryIndex => Option<VerifiableCallOF<T>>
+	#[pallet::storage]
+	#[pallet::getter(fn active_query)]
+	pub(super) type ActiveQuery<T: Config> = StorageMap<_, Identity, QueryIndex, VerifiableCallOf<T>, OptionQuery>;
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Index key does not have an active query currently
+		NoMatchingCall,
+		/// Verifiable Call is too large
+		TooLargeVerifiableCall,
+		/// Query has expired
+		QueryExpired,
+		/// Query has not yet expired
+		QueryNotExpired,
+		/// Not account that requested query
+		NotQueryAccount,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Dispatch task with arbitrary data in origin.
+		///
+		/// - `query_index`: Unique index mapped to a particular query
+		/// - `data`: Aribtrary data to be injected into the call via origin
+		#[pallet::weight(0)]
+		#[transactional]
+		pub fn dispatch_task(origin: OriginFor<T>, query_index: QueryIndex, data: Vec<u8>) -> DispatchResult {
```

Comment:
The name again is a bit confusing. Dispatch a task, for a given query? And we get a query from `ActiveQuery` which is a `VerifiableCall`.

This should be relay_response (answer to a RelayRequest / Query).

Reply:
Renamed to respond_query_request